src/procedures/proc_functions.c

```diff
@@ -0,0 +1,192 @@
+#include "../RG.h"
+#include "procedure.h"
+#include "proc_functions.h"
+#include "../datatypes/array.h"
+
+extern rax *__aeRegisteredFuncs;
+
+// CALL dbms.functions()
+
+typedef struct {
+    SIValue *output;                      // array with a one entry: [name]
+    raxIterator iter;                     // procedures iterator
+    SIValue *yield_name;                  // yield name
+    SIValue *yield_signature;             // yield the function signature in the format name: t1 ... tn -> tn+1
+    SIValue *yield_description;           // yield document string
+    SIValue *yield_return_description;    // yield the description of the return value of the function
+    SIValue *yield_argument_description;  // yield the description of the function arguments
+} ProcFunctionsPrivateData;
+
+static void _process_yield
+(
+	ProcFunctionsPrivateData *ctx,
+	const char **yield
+) {
+    ctx->yield_name                 = NULL;
+    ctx->yield_signature            = NULL;
+    ctx->yield_description          = NULL;
+    ctx->yield_return_description   = NULL;
+    ctx->yield_argument_description = NULL;
+
+    int idx = 0;
+	for(uint i = 0; i < array_len(yield); i++) {
+        if(strcasecmp("name", yield[i]) == 0) {
+            ctx->yield_name = ctx->output + idx;
+            idx++;
+            continue;
+        }
+        if(strcasecmp("signature", yield[i]) == 0) {
+            ctx->yield_signature = ctx->output + idx;
+            idx++;
+            continue;
+        }
+        if(strcasecmp("description", yield[i]) == 0) {
+            ctx->yield_description = ctx->output + idx;
+            idx++;
+            continue;
+        }
+        if(strcasecmp("return_description", yield[i]) == 0) {
+            ctx->yield_return_description = ctx->output + idx;
+            idx++;
+            continue;
+        }
+        if(strcasecmp("argument_description", yield[i]) == 0) {
+            ctx->yield_argument_description = ctx->output + idx;
+            idx++;
+            continue;
+        }
+    }
+}
+
+ProcedureResult Proc_FunctionsInvoke
+(
+    ProcedureCtx *ctx,
+    const SIValue *args,
+    const char **yield
+) {
+    if(array_len((SIValue *)args) != 0) return PROCEDURE_ERR;
+
+    ProcFunctionsPrivateData *pdata = rm_malloc(sizeof(ProcFunctionsPrivateData));
+    memset(pdata, 0, sizeof(ProcFunctionsPrivateData));
+
+    // initialize an iterator to the rax that contains all functions
+    rax *functions = __aeRegisteredFuncs;
+    raxStart(&pdata->iter, functions);
+    raxSeek(&pdata->iter, "^", NULL, 0);
+    pdata->output = array_new(SIValue, 5);
+    _process_yield(pdata, yield);
+
+    ctx->privateData = pdata;
+    return PROCEDURE_OK;
+}
+
+
+// promote the rax iterator to the next function and return its name and mode.
+SIValue *Proc_FunctionsStep
+(
+    ProcedureCtx *ctx
+) {
+    ASSERT(ctx->privateData);
+    ASSERT(ctx->privateData != NULL);
+    
+    size_t bufferLen = MULTIPLE_TYPE_STRING_BUFFER_SIZE * 15;
```

Comment:
I don't understand how did we came up with this size,
For dynamic string construction consider using our `sds` struct